packages/material-ui/src/styles/createTypography.js

```diff
@@ -15,118 +26,229 @@ export default function createTypography(palette, typography) {
+      textTransform: 'uppercase',
+    },
+    caseSentence = {
+      '&:first-letter': {
+        textTransform: 'uppercase',
+      },
+    },
+    ignoreDeprecationWarnings = process.env.MUI_SUPPRESS_DEPRECATION_WARNINGS,
+    useNextVariants = false,
     // Apply the CSS properties to all the variants.
     allVariants,
     ...other
   } = typeof typography === 'function' ? typography(palette) : typography;
 
+  warning(
+    !Object.keys(other).some(variant => deprecatedVariants.includes(variant)),
+    'Deprecation Warning: Material-UI: You are passing a deprecated variant to ' +
+      `createTypography. ${migrationGuideMessage}`,
+  );
+
+  warning(
+    useNextVariants || !Object.keys(other).some(variant => restyledVariants.includes(variant)),
+    'Deprecation Warning: Material-UI: You are passing a variant to createTypography ' +
+      'that will be restyled in the next major release, without indicating that you ' +
+      `are using typography v2 (set \`useNextVariants\` to true. ${migrationGuideMessage}`,
+  );
+
   const coef = fontSize / 14;
-  function pxToRem(value) {
-    return `${(value / htmlFontSize) * coef}rem`;
-  }
+
+  const letterSpacingToEm = (tracking, spSize) => {
+    return `${(tracking / spSize) * coef}em`;
+  };
+
+  const pxToRem = size => {
+    return `${(size / htmlFontSize) * coef}rem`;
+  };
+
+  const getVariant = (variant, localUseNextVariants) => {
+    warning(
+      ignoreDeprecationWarnings || !deprecatedVariants.includes(variant),
+      'Deprecation Warning: Material-UI: You are using the deprecated typography variant ' +
+        `${variant} that will be removed in the next major release. ${migrationGuideMessage}`,
+    );
+
+    const nextVariant = nextVariantMapping(variant);
+
+    // complete v2 switch
+    if (useNextVariants) {
+      return nextVariant;
+    }
+
+    const isRestyledVariant = restyledVariants.includes(variant);
+    // global v1, local v2
+    if (localUseNextVariants) {
+      if (isRestyledVariant) {
+        return `${nextVariant}Next`;
+      }
+      return nextVariant;
+    }
+
+    // v1 => restyle warnings
+    warning(
+      ignoreDeprecationWarnings || !isRestyledVariant,
+      'Deprecation Warning: Material-UI: You are using the typography variant ' +
+        `${variant} which will be restyled in the next major release. ${migrationGuideMessage}`,
+    );
+
+    return variant;
+  };
+
+  const utils = { getVariant, letterSpacingToEm, pxToRem };
+
+  const propertiesForCategory = (weight, size, casing, letterSpacing) => {
+    return {
+      color: palette.text.primary,
+      fontFamily,
+      fontSize: pxToRem(size),
+      fontWeight: weight,
+      letterSpacing: letterSpacingToEm(letterSpacing, size),
+      ...casing,
+      ...allVariants,
+    };
+  };
+
+  /* eslint-disable key-spacing, no-multi-spaces */
+  // prettier-ignore
+  const nextVariants = {
+    headline1:   propertiesForCategory(fontWeightLight,   96, caseSentence, -1.5),
```

Comment:
Their negative letter-spacing is causing issues. For instance when using different fonts. I have been minimizing the value as much as possible. Also, why not using the `em` value directly?

Reply:
I wanted to follow the spec as closely as possible and they use other values that are then converted depending on the system (web, ios, android). The material-web-components uses the same strategy. I would keep the conversion so that we can easily spot differences if something changes or a if a value is wrong.

I mean any property-value combination can cause problems depending on the font-family, no? The letter-spacing is also not negative on my setup. Rember that those values are also converted.

Reply:
I think if somebody is customizing typography by using another font-family then they are also responsible for adjusting letter-spacing if this causes problems. Maybe we can let users pass a custom function to convert letter-spacing?

Reply:
Since the letter-spacing was designed for Roboto we could not apply the letter-spacing if another font-family is applied. Does this sound like a good compromise?

Reply:
> Since the letter-spacing was designed for Roboto we could not apply the letter-spacing if another font-family is applied. Does this sound like a good compromise?

It sounds perfect.

Reply:
By the way the old typography variants were using letter-spacing.